src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java

```diff
@@ -399,7 +426,54 @@ public void removeVirtualSpout(final VirtualSpoutIdentifier virtualSpoutIdentifi
      * @return true when the spout exists, false when it does not.
      */
     public boolean hasVirtualSpout(final VirtualSpoutIdentifier spoutIdentifier) {
-        return getCoordinator().hasVirtualSpout(spoutIdentifier);
+        checkSpoutOpened();
+        return getSpoutMonitor().hasVirtualSpout(spoutIdentifier);
+    }
+
+    /**
+     * Create and start the SpoutMonitor instance.
+     * @param topologyConfig Topology configuration.
+     * @param topologyContext Topology context information.
+     * @param virtualSpoutMessageBus ThreadSafe message bus.
+     * @param metricsRecorder Metric reporter.
+     */
+    private void startSpoutMonitor(
+        final Map<String, Object> topologyConfig,
+        final TopologyContext topologyContext,
+        final VirtualSpoutMessageBus virtualSpoutMessageBus,
+        final MetricsRecorder metricsRecorder) {
+
+        // Prevent more than one instance from being created.
+        if (spoutMonitor != null) {
+            throw new IllegalStateException("Cannot create multiple spout monitor instances!");
+        }
+
+        // Build base name for threads: SpoutId:TaskIndex
+        final String threadContextName =
+            topologyContext.getThisComponentId() + ":" + topologyContext.getThisTaskIndex();
+
+        // Create instance.
+        spoutMonitor = new SpoutMonitor(
+            topologyConfig,
+            threadContextName,
+            virtualSpoutMessageBus,
+            metricsRecorder
+        );
+
+        // Create and name thread.
+        final Thread thread = new Thread(spoutMonitor, "Spout Monitor on " + threadContextName);
+
+        // Mark as a User thread.
+        thread.setDaemon(false);
+
+        // Start it.  Its intended that this thread will loop forever
+        // and as gracefully as possible handle any errors such that it never stops running
+        // unless its passed an interrupt signal or requested to stop.
+        // We have nothing watching this.
+        thread.start();
+
+        // Save reference to thread.
+        spoutMonitorThread = thread;
```

Comment:
Why not just set this when you create the thread? Why the additional reference?

Reply:
Set this where? on SpoutMonitor?  Seems awkward to place it there.  In reality if we don't really care about being 100% sure the thread stops on close, we don't need to keep the reference at all

Reply:
I’m saying why bother with “thread” at all, why not store the Thread directly on spoutMonitorThread